zerver/lib/markdown/__init__.py

```diff
@@ -575,6 +575,52 @@ def run(self, root: Element) -> None:
             video.set("src", get_camo_url(url))
 
 
+class UnnestLinks(markdown.treeprocessors.Treeprocessor):
+    @override
+    def run(self, root: Element) -> None:
+        # Build a map from each element to its parent. This is necessary
+        # because the standard library ElementTree doesn't have a .parent attribute.
+        parent_map = {child: parent for parent in root.iter() for child in parent}
+
+        # Iterate over a static list of all <a> tags in the document.
+        for a_tag in list(root.iter("a")):
+            # Find any <a> tags that are descendants of the current one.
+            nested_a_tags = list(a_tag.iter("a"))
+
+            # The first element of iter('a') is a_tag itself. We only care about
+            # tags nested inside it, so we skip the first element.
+            for nested_a in nested_a_tags[1:]:
+                parent = parent_map.get(nested_a)
+                if parent is None:
+                    # This should not happen in a well-formed tree, but we check just in case.
+                    continue
+
+                # This is the "unwrapping" logic. We need to replace the nested_a
+                # element with its text content. The text needs to be carefully
+                # stitched back into the parent element's text or the tail of
+                # the preceding element.
+
+                # Combine the nested link's text and tail text.
+                text_to_insert = nested_a.text or ""
+                text_to_insert += nested_a.tail or ""
+
+                # Find the position of the nested_a tag within its parent.
+                for i, child in enumerate(parent):
+                    if child is nested_a:
+                        if i == 0:
+                            # If it's the first child, prepend its text to the parent's text.
+                            parent.text = (parent.text or "") + text_to_insert
+                        else:
+                            # Otherwise, append its text to the tail of the previous sibling.
+                            prev_sibling = parent[i - 1]
+                            prev_sibling.tail = (prev_sibling.tail or "") + text_to_insert
+
+                        # Finally, remove the nested_a tag from the tree.
+                        parent.remove(nested_a)
+                        # We found and processed it, so we can break the inner loop.
+                        break
```

Comment:
This is a pretty complicated bit of logic. We'll want to see a proper test suite for it handling the types of corner cases that it's designed to address.

I'm overall a bit skeptical of the approach of generating invalid HTML and then trying to fix it up. Usually, it's best to avoid generating invalid states in the first place. One idea might be to have a subclass or variant of `AtomicString` that does allow certain processing, but I'm skeptical that can be done simply. Best would be to figure out if there's a way to manage the ordering of steps or how linkifiers work to protect against the failure mode in question directly.